cmd/snap-update-ns/change.go

```diff
@@ -344,14 +356,18 @@ func (c *Change) lowLevelPerform(as *Assumptions) error {
 			// Perform the raw unmount operation.
 			if err == nil {
 				err = sysUnmount(c.Entry.Dir, flags)
+				umountOpts, unknownFlags := mount.UnmountFlagsToOpts(flags)
+				if unknownFlags != 0 {
+					umountOpts = append(umountOpts, fmt.Sprintf("%#x", unknownFlags))
+				}
+				logger.Debugf("umount %q %s (error: %v)", c.Entry.Dir, strings.Join(umountOpts, "|"), err)
+				if err != nil {
+					return err
+				}
 			}
 			if err == nil {
```

Comment:
since we re-assigned to err in the block above, and err is shared inside the block and outside of the block, we can combine this block with the one above it, can't we?, then you would have something like:

```go
			if err == nil {
				err = sysUnmount(c.Entry.Dir, flags)
				umountOpts, unknownFlags := mount.UnmountFlagsToOpts(flags)
				if unknownFlags != 0 {
					umountOpts = append(umountOpts, fmt.Sprintf("%#x", unknownFlags))
				}
				logger.Debugf("umount %q %s (error: %v)", c.Entry.Dir, strings.Join(umountOpts, "|"), err)
				if err != nil {
					return err
				}
			}
			as.AddChange(c)
```

Reply:
I need to re-read this after having a coffee. The code sample you pasted is literally the one in the diff now. I must be missing something.

Reply:
sorry the important part of the diff is that you have:

```go
			if err == nil {
				err = sysUnmount(c.Entry.Dir, flags)
				// other stuff
				if err != nil {
					return err
				}
			}
			if err == nil {
				as.AddChange(c)
			}
```
and I'm proposing combining these two if's into a single one like so:

```go
			if err == nil {
				err = sysUnmount(c.Entry.Dir, flags)
				// other stuff
				if err != nil {
					return err
				}
				as.AddChange(c)
			}
```

this works because while we rewrite err in the sysUnmount, we check if it's nil right afterwards, so by the time we leave your first if, if we didn't return from the function we know that `err == nil` and so we can skip the unnecessary check and just add the change. 

Does that make more sense?

Reply:
Ah, yeah. Totally. I'll do this as a follow-up as it's green and can land straight away.

Reply:
A followup sounds good to me